Feat/boto cfn unit tests

major0 opened the pull request:
### What does this PR do?
Implement unit tests for boto_cfn

### Previous Behavior
No unit tests previously existed for boto_cfn

### New Behavior
- Implemented unit tests for boto_cfn
- Fixed bug allowing the supplying of both `template_body` and `template_url` to `boto_cfn.create()` (disallowed by boto/boto3)
- `boto_cfn.delete()` returns a status dictionary on success instead of the raw boto2 response dictionary.
- Various functions now return a status dictionary on failure instead of attempting to simply return a `str` object.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltstack.com/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- N/A Docs
- [X] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [X] Tests written/updated

### Commits signed with GPG?
Yes

major0 committed:
fix: changes generated by pre-commit scripts

major0 committed:
feat: add boot_cfn unit tests

major0 committed:
fix: boto_cfn create/update/validate assert when conflicting args supplied

A few CloudFormation can accept "either" a template_body as part of the
parameter list "or" a URL to a template, but they can not accept both.

waynew reviewed (dismissed):
🎶 🎸 🎶

Ch3LL commented on tests/unit/modules/test_boto_cfn.py:
```diff
@@ -0,0 +1,163 @@
+
+    def test_delete_none_existant_stack_returns_success(self):
+        name = _random_stack_name()
+        self.assertTrue(boto_cfn.delete(name, **self.conn_parameters))
```
Should we be failing if we are trying to delete something that does not exist? If so, we can try to handle this outside of this PR. I can create an issue if this is the case.

major0 replied:
I am not 100% certain here.  Some of the other boto modules return true when deleting something that does not exist (or creating something that already exists), so I was mostly just mimic'ing that behavior.

Sitting here looking through some of the other modules and it unfortunately looks like a mixed bag.

major0 replied:
I was just looking at `boto_vpc` and noticed that it returns does return `False` when attempting to delete something that does not exist, but it returns it as part of a dictionary such that it can return a status _and_ a message. Should this be the preferred approach for these modules?

major0 replied:
It looks like `boto_iam` is the primary source which returns true if the target resource already exists:
```
def create_instance_profile(name, region=None, key=None, keyid=None, profile=None):
    """
    Create an instance profile.

    CLI Example:

    .. code-block:: bash

        salt myminion boto_iam.create_instance_profile myiprofile
    """
    conn = _get_conn(region=region, key=key, keyid=keyid, profile=profile)

    if instance_profile_exists(name, region, key, keyid, profile):
        return True
    try:
        # This call returns an instance profile if successful and an exception
        # if not. It's annoying.
        conn.create_instance_profile(name)
        log.info("Created %s instance profile.", name)
    except boto.exception.BotoServerError as e:
        log.debug(e)
        log.error("Failed to create %s instance profile.", name)
        return False
    return True
```

IMHO I prefer the approach that boto_vpc takes with returning false on these attempts, but also returning an error message with it.

Ch3LL replied:
Yeah I agree with your approach as well, lthough this would need to be handled with great care since the state modules might not be accounting for this change. ping @waynew thoughts? Do you agree or should we keep as is?

waynew replied:
I think I don't know enough to make a *specific* recommendation.

Here's what I /do/ know:

- Some things in Salt return True when a thing is already gone.
- Some things in Salt return True (with `'changes':`) when something is already gone.
- Some things in Salt return False when they can't do something they wanted to do.

As I understand it, the question is basically one of intent. If the "universe" is in the desired state, then the result is True, regardless of how we got there. If the universe is *not* in the desired state, then the result should be False.

From what little I understand at the moment it looks like the /desired/ state is that the thing is deleted and gone. If that's the case, then whether it was deleted with fire, or we tried to delete it and it was already gone, the universe looks like we desire it, and it should return True.


But I could be wrong! 🙃  (but hopefully the reasoning behind it is clear enough - let me know if not!)

major0 replied:
@waynew That seems to align with the code in question.  In particular some code has an explicit test to return true if the desired state already exists (since the boto layer might return an error otherwise), and some code does not have a test (though still returns true because the AWS boto layer does not error). For example, the AWS `boto.create_stack()` call returns true even if the stack already exists, where the similar AWS boto IAM call will throw an error if you true to create a user that already exists (so `boto_iam.create_user()` performs extra tests).

While digging deeper into this, I realized that nearly all of the boto_cfn code simply returns a `str` on error (taken from the BotoServerError), which means it wasn't possible to test for an error based on truthyness. Further more, `boto_cfn.delete()` returned the boto response dictionary on success (i.e. `return {'DeleteStackResponse': {'DeleteStackResult': {}}}`), and a `str` on error.

major0 committed:
fix: consistent error reporting in boto_cfn

Changed the following module functions to return `{"status": False, "error": str(e)}` on failure:

- `boto_cfn.describe()` instead of `True` when stackname exists but boto fails to return a description
- `boto_cfn.describe()` instead of returning `False` an error
- `boto_cfn.create()` instead of `False` on error
- `boto_cfn.delete()` instead of returning `str(e)` on error
- `boto_cfn.update_stack()` instead of returning `str(e)` on error
- `boto_cfn.et_template()` instead of returning `str(e)` on error

major0 committed:
fix: change `boto_cfn.describe()` to include status on success.

Previously returned `{"stack": data}`

Now returns `{"stack": data, "status", True}`

major0 committed:
fix: change various `boto_cfn` return responses on success

Changed the following module functions to include `"status" = True` in a results dictionary on success:
- `boto_cfn.describe()` previously returned a desciption dictionary w/out status data
- `boto_cfn.delete()` previously returned the raw boto response dictionary with no data

major0 committed:
Merge branch 'master' into feat/boto_cfn_unit_tests

dwoz commented:
Closing this due to inactivity. Anyone should feel free to re-open it if they want to see it through to the end in one release cycle.
